### PR TITLE
Upload frontend strings to crowdin

### PR DESCRIPTION
**Pull request by mbenadda**
## Purpose

Generate a POT file from our strings and upload strings to our SaaS translation manager.

## Proposal

- [x] tidy up our folder structure
- [x] use `react-intl-po` to generate a standard POT from our JSON files with message descriptors
- [x] write a small node script to handle string uploading
- [x] add a Circle CI job to upload strings when running on master

🚫 install & configure the Crowdin CLI tool (abandoned because it introduced a dependency to java in the project and that's not something we're ready for)

Note: penultimate step towards #89

### CODE REVIEW

File: .babelrc.js [hunk 1]
```diff
@@ -3,7 +3,7 @@ module.exports = {
     [
       'react-intl',
       {
-        messagesDir: './strings/',
+        messagesDir: './i18n/strings',
```
Comment (jmaupetit) [inline .babelrc.js]: Is the `strings` directory required?

File: docker/images/dev/Dockerfile [hunk 2]
```diff
@@ -16,6 +16,11 @@ RUN apt-get update && \
     vim && \
     rm -rf /var/lib/apt/lists/*
 
+# Install the Crowdin CLI that helps us upload/download strings & translations
+RUN wget -qO - https://artifacts.crowdin.com/repo/GPG-KEY-crowdin | apt-key add - && \
+    echo "deb https://artifacts.crowdin.com/repo/deb/ /" > /etc/apt/sources.list.d/crowdin.list && \
+    apt-get install crowdin
```
Comment (jmaupetit) [inline docker/images/dev/Dockerfile]: You should update package list prior to installation, _e.g_ `apt-get update && apt-get install crowdin`

Comment (jmaupetit) [reply docker/images/dev/Dockerfile]: Also clear apt cache to decrease image size by adding: `rm -rf /var/lib/apt/lists/*` as the final instruction of this layer (`RUN`).

Comment (jmaupetit) [reply docker/images/dev/Dockerfile]: :thinking: After-thought: maybe we can improve the two RUN statements using apt as follow:

```Docker
# Add Crowdin CLI (upload/download strings & translations) apt repository
RUN wget -qO - https://artifacts.crowdin.com/repo/GPG-KEY-crowdin | apt-key add - && \
    echo "deb https://artifacts.crowdin.com/repo/deb/ /" > /etc/apt/sources.list.d/crowdin.list 

# Install development dependencies via apt
RUN apt-get update && \
    apt-get install -y \
      crowdin vim && \
    rm -rf /var/lib/apt/lists/*
```

File: i18n/scripts/upload.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            `🙉 cannot update ${fileName}, attempting to create instead...`,
+          );
+          return createOrUpdateFile(fileName, filePath, true);
+        }
+
+        console.log(`💥 failed to upload ${fileName}, aborting.`);
+        process.exit(1);
+      },
+    );
+}
```
Comment (jmaupetit) [inline i18n/scripts/upload.js]: Is the default exit code 0 for a node script?

Comment (mbenadda) [reply i18n/scripts/upload.js]: It is

**Comment by mbenadda** [comment]
This PR is now ready for review.

I did not handle backend strings yet as I'm not sure of the right way to do it: in a regular project I'd need to call `manage.py` but here it feels wrong to make our build dependent on sandbox stuff.

What do you think?

**Comment by sampaccoud** [comment]
@mbenadda I just pushed a POC of how it could work on the backend. Read the commit message for design decisions.

TODO: install gettext on Alpine

File: Makefile [hunk 4]
```diff
@@ -61,6 +61,9 @@ build-ts: ## build TypeScript application
 	@$(YARN) build
 .PHONY: build-ts
 
+compiledmessages: ## compile the gettext files
```
Comment (jmaupetit) [inline Makefile]: Why not calling this rule `compilemessages` to stick with Django management commands?

File: Makefile [hunk 5]
```diff
@@ -100,6 +103,9 @@ logs: ## get development logs
 	@$(COMPOSE) logs -f
 .PHONY: logs
 
+messages: ## create the .po files used for i18n
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply Makefile]: but it is when you run "make messages" ;-)
To be honest, I copied the command name from django Oscar.

File: Makefile (hunk 4 above)
Comment (sampaccoud) [reply Makefile]: True this one. @mbenadda I let you change so we don't step on each other's commits?

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmaupetit) [reply Makefile]: :wink:

File: Makefile (hunk 4 above)
Comment (mbenadda) [reply Makefile]: Yeah I'll do it, I'm updating the PR progressively through several rounds of `rebase -i` so it's definitely better if we don't do it concomitantly

File: .circleci/config.yml [hunk 6]
```diff
@@ -695,13 +745,30 @@ workflows:
...
+      # i18n jobs
+      #
+      # Extract strings and upload them to our translation management SaaS
+      - upload-i18n-strings:
+          requires:
+            - build-front
+            - extract-back-strings
+          filters:
+            tags:
+              only: /.*/ # master
```
Comment (mbenadda) [inline .circleci/config.yml]: Note: change this before merging

File: .circleci/config.yml [hunk 7]
```diff
@@ -695,13 +745,30 @@ workflows:
           filters:
             tags:
               only: /.*/
+      - extract-back-strings:
+          requires:
+            - build-dev
+          filters:
+            tags:
+              only: /.*/ # master
```
Comment (mbenadda) [inline .circleci/config.yml]: Note: change this before merging

File: .circleci/config.yml [hunk 8]
```diff
@@ -258,6 +258,46 @@ jobs:
...
+      - restore_cache:
+          keys:
+          - v3-front-dependencies-{{ checksum "package.json" }}
+      - attach_workspace:
+          at: ~/fun/src/richie/locale
+      - run:
+          name: Use our node script to upload i18n backend strings file
+          command: node i18n/scripts/upload.js 'richie-back.pot' '/../../../fun/src/richie/locale/django.pot'
+      - attach_workspace:
+          at: ~/richie/src/richie-front/i18n
```
Comment (jmaupetit) [inline .circleci/config.yml]: We need more consistency in our paths (_i.e._ `~/richie`, `~/fun`), but this is not the purpose of this work...

File: .circleci/config.yml [hunk 9]
```diff
@@ -258,6 +258,46 @@ jobs:
...
+      - restore_cache:
+          keys:
+            - docker-debian-images-dev-{{ .Revision }}
+      - run:
+          name: Load images to docker engine
+          command: |
+            docker load < docker/images/dev/richie.tar
+      - run:
+          name: Generate a POT file from strings extracted from the project
+          command: make messages
```
Comment (jmaupetit) [inline .circleci/config.yml]: :warning: Using this `Makefile`'s rule, you'll use the development's `docker-compose` configuration. This is not what you want. Instead you can run the `makemessages` management command using the current docker image:

```
docker run --rm \
    --user="$(id -u)" \
    --volume "$PWD:/app" \
    --workdir /app/src/richie \
    "richie:${CIRCLE_SHA1}-dev" \
        python /app/sandbox/manage.py makemessages --keep-pot
```

**Comment by mbenadda** [comment]
I think this PR is ready to be reviewed/merged. Thanks for your help @jmaupetit @sampaccoud 👍

File: .circleci/config.yml [hunk 10]
```diff
@@ -258,6 +258,52 @@ jobs:
               run --rm --no-deps app \
                 black src/richie/apps src/richie/plugins sandbox --check
 
+  extract-back-strings:
```
Comment (sampaccoud) [inline .circleci/config.yml]: we usualy add docstrings on top of the jobs with all the insight about what you've done and why you've done it like this.

Is there a reason why you didn't do it in the back's "build" job like you did for the front?

File: .gitignore [hunk 11]
```diff
@@ -34,6 +34,7 @@ nosetests.xml
 coverage.xml
 
 # Translations
+*.pot
 *.mo
```
Comment (sampaccoud) [inline .gitignore]: I think we will need to remove .mo files from the .gitignore if we want to commit translations. WDYT @jmaupetit?

File: .gitignore [hunk 12]
```diff
@@ -81,3 +82,7 @@ node_modules
 
 # logs
 yarn-error.log
+
+# Exported strings & generated translation template
+src/richie-front/i18n/src
+src/richie-front/i18n/richie-front.pot
```
Comment (sampaccoud) [inline .gitignore]: this should be covered by *.pot

File: .gitignore (hunk 11 above)
Comment (mbenadda) [reply .gitignore]: Will we not commit the `.po` files? I think `.mo` is a non-human-readable format that can be produced from `.po` and therefore not useful to commit?

File: .circleci/config.yml (hunk 10 above)
Comment (mbenadda) [reply .circleci/config.yml]: As far as I can see there is no backend "build" job

File: .gitignore (hunk 11 above)
Comment (sampaccoud) [reply .gitignore]: yes. But then we need to add compilemessages to the build in circleci ;-)

File: .circleci/config.yml (hunk 10 above)
Comment (sampaccoud) [reply .circleci/config.yml]: There are 3 of them: build, build-dev and build-alpine.

I agree that the names are not coherent. We should probably rename them to: build-back-debian, build-back-debian-dev and build-back-alpine...

Comment (jmaupetit) [reply .circleci/config.yml]: :+1:

File: .gitignore (hunk 11 above)
Comment (jmaupetit) [reply .gitignore]: Yup. `.mo` files should be in the repository and bundled in the distributed image/package.

**Pull request merged**